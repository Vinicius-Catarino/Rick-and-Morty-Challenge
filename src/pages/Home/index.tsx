import React, { useCallback, useEffect, useState } from "react";

import Loading from "../../components/Loading";
import CardCharacter from "../../components/CardCharacter";
import ModalCharacterDetails from "../../components/ModalCharacterDetails";

import Pagination from "@material-ui/lab/Pagination";

import Rickandmorty from "../../assets/Rickandmorty.png";

import getAllCharacters from "../../services/graphql/getAllCharacters";

import Character from "../../models/Character";

import {
  Container,
  ContainerSearch,
  ContainerCard,
  ContainerPagination,
  Text,
} from "./styles";
import getCharactersByFilter from "../../services/graphql/getCharactersByFilter";

const Home: React.FC = () => {
  const [isLoading, setIsLoading] = useState(false);
  const [characterToSearch, setCharacterToSearch] = useState("");
  const [characters, setCharacters] = useState<Character[]>([]);
  const [selectedCharacter, setSelectedCharacter] = useState<Character>();
  const [openModalSelected, setOpenModalSelected] = useState(false);
  const [paginationNumber, setPaginationNumber] = useState(1);
  const [countCharacters, setCountCharacters] = useState(0);
  const [errorText, setErrorText] = useState("");

  const loadAllCharacters = useCallback(async (page) => {
    try {
      setIsLoading(true);

      const response = await getAllCharacters(page);

      setCountCharacters(response.info.count);
      setCharacters(response.results);
      setPaginationNumber(page);
      setIsLoading(false);
    } catch (error) {
      setErrorText("An error occurred while fetching the characters");
      setIsLoading(false);
    }
  }, []);

  useEffect(() => {
    loadAllCharacters(paginationNumber);
  }, [paginationNumber, loadAllCharacters]);

  const handleCharacterSearch = useCallback(async () => {
    if (!characterToSearch) {
      return setErrorText("Please enter a valid name");
    }

    try {
      setIsLoading(true);
      setPaginationNumber(1);

      const response = await getCharactersByFilter(
        paginationNumber,
        characterToSearch
      );

      if (!response) setErrorText("No character with this name");
      else setErrorText("");

      setCharacters(response.results);
      setIsLoading(false);
    } catch (error) {
      setErrorText("An error occurred while fetching the characters");
      setIsLoading(false);
    }
  }, [characterToSearch, paginationNumber]);

  const handleModal = useCallback(() => {
    setOpenModalSelected(!openModalSelected);
  }, [openModalSelected]);

  return (
    <>
      <Container isLoading={isLoading}>
        <ContainerSearch>
          <img src={Rickandmorty} alt="Rick and Morty" />
          <div>
            <input
              onKeyPress={(e) => {
                if (e.key === "Enter") handleCharacterSearch();
              }}
              id="inputTextToSearch"
              value={characterToSearch}
              onChange={(e) => setCharacterToSearch(e.target.value)}
              placeholder="Search character"
            />
            <button onClick={() => handleCharacterSearch()} id="buttonSearch">
              Search
            </button>
          </div>
          <Text>{errorText}</Text>
        </ContainerSearch>

        <ContainerCard>
          {characters.map((character) => (
            <CardCharacter
              onClick={() => {
                setSelectedCharacter(character);
                handleModal();
              }}
              key={character.id}
              character={character}
            />
          ))}
        </ContainerCard>

        <ContainerPagination>
          <Pagination
            count={countCharacters / 20}
            onChange={(e, page) => loadAllCharacters(page)}
            color="primary"
          />
        </ContainerPagination>

        {selectedCharacter && (
          <ModalCharacterDetails
            open={openModalSelected}
            character={selectedCharacter}
            handleClose={() => handleModal()}
          />
        )}
      </Container>
      {isLoading && <Loading />}
    </>
  );
};

export default Home;
