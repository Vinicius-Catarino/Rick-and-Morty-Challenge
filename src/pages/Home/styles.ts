import styled, { css } from "styled-components";

import BackgroundImage from "../../assets/Background.png";

interface IContainerProps {
  isLoading: boolean;
}

export const Container = styled.div<IContainerProps>`
  justify-content: center;
  align-items: center;

  overflow: auto;
  min-height: 100%;
  min-width: 100%;

  width: 100%;
  height: 100%;

  padding: 0px;

  position: fixed;
  top: 0;
  left: 0;

  background: url(${BackgroundImage}) no-repeat center center fixed;
  background-size: cover;

  ${(props) =>
    props.isLoading &&
    css`
      filter: blur(12px);
    `}
`;

export const Form = styled.form`
  display: flex;
  justify-content: center;
  align-items: center;
  flex-direction: column;

  div {
    margin-top: 60px;
  }

  input {
    border: 2px solid #ffffff;
    box-sizing: border-box;
    border-radius: 8px;
    background: transparent;
    padding: 6px 8px;
    margin-right: 16px;
    color: #9f9f9f;
  }

  button {
    padding: 6px 8px;
    border: 2px solid #ffffff;
    box-sizing: border-box;
    border-radius: 8px;
    background: transparent;
    color: #fff;
  }
`;

export const ContainerCard = styled.div`
  display: flex;
  flex-wrap: wrap;
  max-width: 960px;
  margin: 0 auto;
  padding: 40px;
  justify-content: center;
`;

export const ContainerPagination = styled.div`
  display: flex;
  justify-content: center;
  margin: 20px auto;
`;
