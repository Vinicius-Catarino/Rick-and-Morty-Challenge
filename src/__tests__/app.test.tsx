import React from "react";
import { shallow } from "enzyme";
import Root from "../App";
import Home from "../pages/Home";
import "jsdom-global/register";
import axios from "axios";
// import getCharactersByFilter from "../services/graphql/getCharactersByFilter";

// jest.mock("../services/graphql/getCharactersByFilter");

const charactersReturn = {
  results: [
    {
      id: "1",
      name: "Rick Sanchez",
      type: "",
      image: "https://rickandmortyapi.com/api/character/avatar/1.jpeg",
      species: "Human",
      location: {
        id: "20",
        name: "Earth (Replacement Dimension)",
        type: "Planet",
        dimension: "Replacement Dimension",
      },
      origin: {
        id: "1",
        name: "Earth (C-137)",
        type: "Planet",
        dimension: "Dimension C-137",
      },
    },
  ],
};

describe("index page", () => {
  it("should have Root component", () => {
    const subject: any = shallow(<Root />);
    expect(subject).toMatchSnapshot();
  });

  it("should not be able search without text", (done) => {
    const wrapper = shallow(<Home />);

    const inputTextToSearch = wrapper.find("#inputTextToSearch");

    inputTextToSearch.simulate("change", { target: { value: "" } });

    const button = wrapper.find("#buttonSearch");

    // Auth.signIn = jest.fn().mockImplementation(() => mockUser);

    button.simulate("click");

    expect(wrapper).toMatchSnapshot();
    done();
  });

  it("should be able search", async (done) => {
    const wrapper = shallow(<Home />);

    const inputTextToSearch = wrapper.find("#inputTextToSearch");

    inputTextToSearch.simulate("change", { target: { value: "rick" } });

    const button = wrapper.find("#buttonSearch");

    axios.post = jest.fn().mockImplementation(() => charactersReturn);

    await button.simulate("click");

    setTimeout(() => {
      expect(wrapper).toMatchSnapshot();
      done();
    }, 2000);
  });

  it("should be able search with error", async (done) => {
    const wrapper = shallow(<Home />);

    const inputTextToSearch = wrapper.find("#inputTextToSearch");

    inputTextToSearch.simulate("change", { target: { value: "rick" } });

    const button = wrapper.find("#buttonSearch");

    axios.post = jest.fn().mockImplementation(() => {
      throw new Error("Error");
    });

    await button.simulate("click");

    setTimeout(() => {
      expect(wrapper).toMatchSnapshot();
      done();
    }, 2000);
  });
});
