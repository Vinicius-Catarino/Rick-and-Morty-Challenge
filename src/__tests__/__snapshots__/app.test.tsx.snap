// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`index page should be able search 1`] = `
<Fragment>
  <styled.div
    isLoading={false}
  >
    <styled.form>
      <img
        alt="Rick and Morty"
        src={null}
      />
      <div>
        <input
          id="inputTextToSearch"
          onChange={[Function]}
          placeholder="Search character"
          value="rick"
        />
        <button
          id="buttonSearch"
          onClick={[Function]}
        >
          Search
        </button>
      </div>
    </styled.form>
    <styled.div>
      <CardCharacter
        img="https://rickandmortyapi.com/api/character/avatar/1.jpeg"
        key="1"
        name="Rick Sanchez"
        onClick={[Function]}
        type="Human"
      />
      <CardCharacter
        img="https://rickandmortyapi.com/api/character/avatar/8.jpeg"
        key="8"
        name="Adjudicator Rick"
        onClick={[Function]}
        type="Human"
      />
      <CardCharacter
        img="https://rickandmortyapi.com/api/character/avatar/15.jpeg"
        key="15"
        name="Alien Rick"
        onClick={[Function]}
        type="Alien"
      />
      <CardCharacter
        img="https://rickandmortyapi.com/api/character/avatar/19.jpeg"
        key="19"
        name="Antenna Rick"
        onClick={[Function]}
        type="Human"
      />
      <CardCharacter
        img="https://rickandmortyapi.com/api/character/avatar/22.jpeg"
        key="22"
        name="Aqua Rick"
        onClick={[Function]}
        type="Humanoid"
      />
      <CardCharacter
        img="https://rickandmortyapi.com/api/character/avatar/48.jpeg"
        key="48"
        name="Black Rick"
        onClick={[Function]}
        type="Human"
      />
      <CardCharacter
        img="https://rickandmortyapi.com/api/character/avatar/56.jpeg"
        key="56"
        name="Bootleg Portal Chemist Rick"
        onClick={[Function]}
        type="Human"
      />
      <CardCharacter
        img="https://rickandmortyapi.com/api/character/avatar/69.jpeg"
        key="69"
        name="Commander Rick"
        onClick={[Function]}
        type="Human"
      />
      <CardCharacter
        img="https://rickandmortyapi.com/api/character/avatar/72.jpeg"
        key="72"
        name="Cool Rick"
        onClick={[Function]}
        type="Human"
      />
      <CardCharacter
        img="https://rickandmortyapi.com/api/character/avatar/74.jpeg"
        key="74"
        name="Cop Rick"
        onClick={[Function]}
        type="Human"
      />
      <CardCharacter
        img="https://rickandmortyapi.com/api/character/avatar/78.jpeg"
        key="78"
        name="Cowboy Rick"
        onClick={[Function]}
        type="Human"
      />
      <CardCharacter
        img="https://rickandmortyapi.com/api/character/avatar/82.jpeg"
        key="82"
        name="Cronenberg Rick"
        onClick={[Function]}
        type="Cronenberg"
      />
      <CardCharacter
        img="https://rickandmortyapi.com/api/character/avatar/86.jpeg"
        key="86"
        name="Cyclops Rick"
        onClick={[Function]}
        type="Humanoid"
      />
      <CardCharacter
        img="https://rickandmortyapi.com/api/character/avatar/103.jpeg"
        key="103"
        name="Doofus Rick"
        onClick={[Function]}
        type="Human"
      />
      <CardCharacter
        img="https://rickandmortyapi.com/api/character/avatar/119.jpeg"
        key="119"
        name="Evil Rick"
        onClick={[Function]}
        type="Humanoid"
      />
      <CardCharacter
        img="https://rickandmortyapi.com/api/character/avatar/135.jpeg"
        key="135"
        name="Garment District Rick"
        onClick={[Function]}
        type="Human"
      />
      <CardCharacter
        img="https://rickandmortyapi.com/api/character/avatar/164.jpeg"
        key="164"
        name="Insurance Rick"
        onClick={[Function]}
        type="Human"
      />
      <CardCharacter
        img="https://rickandmortyapi.com/api/character/avatar/165.jpeg"
        key="165"
        name="Investigator Rick"
        onClick={[Function]}
        type="Human"
      />
      <CardCharacter
        img="https://rickandmortyapi.com/api/character/avatar/187.jpeg"
        key="187"
        name="Juggling Rick"
        onClick={[Function]}
        type="Human"
      />
      <CardCharacter
        img="https://rickandmortyapi.com/api/character/avatar/215.jpeg"
        key="215"
        name="Maximums Rickimus"
        onClick={[Function]}
        type="Human"
      />
    </styled.div>
    <styled.div>
      <WithStyles(ForwardRef(Pagination))
        color="primary"
        count={0}
        onChange={[Function]}
      />
    </styled.div>
  </styled.div>
</Fragment>
`;

exports[`index page should be able search with error 1`] = `
<Fragment>
  <styled.div
    isLoading={false}
  >
    <styled.form>
      <img
        alt="Rick and Morty"
        src={null}
      />
      <div>
        <input
          id="inputTextToSearch"
          onChange={[Function]}
          placeholder="Search character"
          value="rick"
        />
        <button
          id="buttonSearch"
          onClick={[Function]}
        >
          Search
        </button>
      </div>
    </styled.form>
    <styled.div>
      <CardCharacter
        img="https://rickandmortyapi.com/api/character/avatar/1.jpeg"
        key="1"
        name="Rick Sanchez"
        onClick={[Function]}
        type="Human"
      />
      <CardCharacter
        img="https://rickandmortyapi.com/api/character/avatar/8.jpeg"
        key="8"
        name="Adjudicator Rick"
        onClick={[Function]}
        type="Human"
      />
      <CardCharacter
        img="https://rickandmortyapi.com/api/character/avatar/15.jpeg"
        key="15"
        name="Alien Rick"
        onClick={[Function]}
        type="Alien"
      />
      <CardCharacter
        img="https://rickandmortyapi.com/api/character/avatar/19.jpeg"
        key="19"
        name="Antenna Rick"
        onClick={[Function]}
        type="Human"
      />
      <CardCharacter
        img="https://rickandmortyapi.com/api/character/avatar/22.jpeg"
        key="22"
        name="Aqua Rick"
        onClick={[Function]}
        type="Humanoid"
      />
      <CardCharacter
        img="https://rickandmortyapi.com/api/character/avatar/48.jpeg"
        key="48"
        name="Black Rick"
        onClick={[Function]}
        type="Human"
      />
      <CardCharacter
        img="https://rickandmortyapi.com/api/character/avatar/56.jpeg"
        key="56"
        name="Bootleg Portal Chemist Rick"
        onClick={[Function]}
        type="Human"
      />
      <CardCharacter
        img="https://rickandmortyapi.com/api/character/avatar/69.jpeg"
        key="69"
        name="Commander Rick"
        onClick={[Function]}
        type="Human"
      />
      <CardCharacter
        img="https://rickandmortyapi.com/api/character/avatar/72.jpeg"
        key="72"
        name="Cool Rick"
        onClick={[Function]}
        type="Human"
      />
      <CardCharacter
        img="https://rickandmortyapi.com/api/character/avatar/74.jpeg"
        key="74"
        name="Cop Rick"
        onClick={[Function]}
        type="Human"
      />
      <CardCharacter
        img="https://rickandmortyapi.com/api/character/avatar/78.jpeg"
        key="78"
        name="Cowboy Rick"
        onClick={[Function]}
        type="Human"
      />
      <CardCharacter
        img="https://rickandmortyapi.com/api/character/avatar/82.jpeg"
        key="82"
        name="Cronenberg Rick"
        onClick={[Function]}
        type="Cronenberg"
      />
      <CardCharacter
        img="https://rickandmortyapi.com/api/character/avatar/86.jpeg"
        key="86"
        name="Cyclops Rick"
        onClick={[Function]}
        type="Humanoid"
      />
      <CardCharacter
        img="https://rickandmortyapi.com/api/character/avatar/103.jpeg"
        key="103"
        name="Doofus Rick"
        onClick={[Function]}
        type="Human"
      />
      <CardCharacter
        img="https://rickandmortyapi.com/api/character/avatar/119.jpeg"
        key="119"
        name="Evil Rick"
        onClick={[Function]}
        type="Humanoid"
      />
      <CardCharacter
        img="https://rickandmortyapi.com/api/character/avatar/135.jpeg"
        key="135"
        name="Garment District Rick"
        onClick={[Function]}
        type="Human"
      />
      <CardCharacter
        img="https://rickandmortyapi.com/api/character/avatar/164.jpeg"
        key="164"
        name="Insurance Rick"
        onClick={[Function]}
        type="Human"
      />
      <CardCharacter
        img="https://rickandmortyapi.com/api/character/avatar/165.jpeg"
        key="165"
        name="Investigator Rick"
        onClick={[Function]}
        type="Human"
      />
      <CardCharacter
        img="https://rickandmortyapi.com/api/character/avatar/187.jpeg"
        key="187"
        name="Juggling Rick"
        onClick={[Function]}
        type="Human"
      />
      <CardCharacter
        img="https://rickandmortyapi.com/api/character/avatar/215.jpeg"
        key="215"
        name="Maximums Rickimus"
        onClick={[Function]}
        type="Human"
      />
    </styled.div>
    <styled.div>
      <WithStyles(ForwardRef(Pagination))
        color="primary"
        count={0}
        onChange={[Function]}
      />
    </styled.div>
  </styled.div>
</Fragment>
`;

exports[`index page should have Root component 1`] = `
<Fragment>
  <BrowserRouter>
    <Routes />
    <Memo(l) />
  </BrowserRouter>
</Fragment>
`;

exports[`index page should not be able search without text 1`] = `
<Fragment>
  <styled.div
    isLoading={false}
  >
    <styled.form>
      <img
        alt="Rick and Morty"
        src={null}
      />
      <div>
        <input
          id="inputTextToSearch"
          onChange={[Function]}
          placeholder="Search character"
          value=""
        />
        <button
          id="buttonSearch"
          onClick={[Function]}
        >
          Search
        </button>
      </div>
    </styled.form>
    <styled.div />
    <styled.div>
      <WithStyles(ForwardRef(Pagination))
        color="primary"
        count={0}
        onChange={[Function]}
      />
    </styled.div>
  </styled.div>
</Fragment>
`;
